src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -163,5 +176,107 @@ private IEnumerable<CodeRefactoringProvider> GetProviders(Document document)
 
             return null;
         }
+
+        private ImmutableArray<CodeRefactoringProvider> GetProjectRefactorings(Project project)
+        {
+            // TODO (https://github.com/dotnet/roslyn/issues/4932): Don't restrict refactorings in Interactive
+            return project.Solution.Workspace.Kind == WorkspaceKind.Interactive
+                ? ImmutableArray<CodeRefactoringProvider>.Empty
+                : _projectRefactoringsMap.GetValue(project.AnalyzerReferences, pId => new StrongBox<ImmutableArray<CodeRefactoringProvider>>(ComputeProjectRefactorings(project))).Value;
```

Comment:
pity that this will always allocate a closure due to 'project' being captured.

Reply:
`this` is also captured. We can possibly avoid allocations if we refactor the code a bit:

```csharp
: _projectRefactoringsMap.TryGetValue (project.AnalyzerReferences, out var value)
? value
: ComputeRefactoringMap(project);

ImmutableArray<CodeRefactoringProvider> ComputeRefactoringMap(project) =>
    _projectRefactoringsMap.GetValue(project.AnalyzerReferences, pId => new StrongBox<ImmutableArray<CodeRefactoringProvider>>(ComputeProjectRefactorings(project))).Value;
```

Reply:
Pattern is used [1](https://github.com/dotnet/roslyn/blob/cfd69157e42e7a082699193d7c2f3290ce0ebb63/src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex.cs#L65) [2](https://github.com/dotnet/roslyn/blob/c64963b60e28c8f3577c5424571656656308f066/src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs#L40) other places

Reply:
Captures only appear on slow paths now.